proposals/type-classes.md

```diff
@@ -0,0 +1,300 @@
+
+```kotlin
+instance class OptionMonoid<A>(instance MA: Monoid<A>) : Monoid<Option<A>> //class position
+
+fun <A> add(a: A, b: A, instance MA: Monoid<A>): A = a.combine(b) //function position
+```
+
+## Compiler Changes
+
+- The type checker will declare the below definition as valid since the `given` clause provides evidence that call sites won't be able to compile calls to this function unless a `Monoid<A>` is in scope.
+```kotlin
+fun <A> add(a: A, b: A): A given Monoid<A> = a.combine(b) //compiles
+```
+- The type checker will declare the below definition as invalid since there is no `Monoid<Int>` in scope.
+```kotlin
+add(1, 2)
+```
+- The type checker will declare the below definition as valid since there is a `Monoid<Int>` in scope.
+```kotlin
+import intext.IntMonoid
+add(1, 2)
+```
+- The type checker will declare the below definition as valid since there is a `Monoid<Int>` in scope.
+```kotlin
+fun addInts(a: Int, b: Int): Int given Monoid<Int> = add(a, b)
+```
+- The type checker will declare the below definition as valid since there is a `with` block around the concrete `IntMonoid` in scope.
+```kotlin
+fun addInts(a: Int, b: Int): Int = with(IntMonoid) { add(a, b) }
+```
+
+## Compile resolution rules
+
+When the compiler finds a call site invoking a function that has type class instances constrains declared with `given` as in the example below:
+
+Declaration: 
+```kotlin
+fun <A> add(a: A, b: A): A given Monoid<A> = a.combine(b) 
+```
+Call site:
+```kotlin
+instance class AddingInts {
+  fun addInts(): Int = add(1, 2)
+}
+```
+The compiler may choose the following order for resolving the evidence that a `Monoid<Int>` exists in scope.
+
+1. Look in the most immediate scope for declarations of `given Monoid<Int>` in this case the function `addInts`
+
+This will compile because the responsibility of providing `Monoid<Int>` is passed unto the callers of `addInts()`:
+```kotlin
+instance class AddingInts {
+  fun addInts(): Int given Monoid<Int> = add(1, 2)
+}
+```
+
+2. Look in the immediately outher class/interface scope for declarations of `given Monoid<Int>` in this case the class `AddingInts`:
+```kotlin
+instance class AddingInts given Monoid<Int> {
+  fun addInts(): Int = add(1, 2)
+}
+```
+This will compile because the responsibility of providing `Monoid<Int>` is passed unto the callers of `AddingInts()`
+
+3. Look in the import declarations for an explicitly imported instance that satisfies the constrain `Monoid<Int>`:
+```kotlin
+import intext.IntMonoid
+instance class AddingInts {
+  fun addInts(): Int = add(1, 2)
+}
+```
+This will compile because the responsibility of providing `Monoid<Int>` is satisfied by `import intext.IntMonoid`
+
+4. Fail to compile if neither outer scopes nor explicit imports fail to provide evidence that there is a constrain satisfied by an instance in scope.
+```kotlin
+import intext.IntMonoid
```

Comment:
The example 4 is the same of 3, should you remove this import?

Reply:
@fvasco feel free to submit a PR with those changes against our fork and we can get them integrated in the proposal.